contracts/bonding/BondingManager.sol

```diff
@@ -885,20 +805,105 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
+            uint256 newStake = transcoderTotalStake(_delegate).sub(_amount);
+            uint256 nextRound = roundsManager().currentRound().add(1);
 
+            transcoderPool.updateKey(_delegate, newStake, address(0), address(0));
+            nextRoundTotalActiveStake = nextRoundTotalActiveStake.sub(_amount);
+            Transcoder storage t = transcoders[_delegate];
+            t.lastActiveStakeUpdateRound = nextRound;
+            t.earningsPoolPerRound[nextRound].setStake(newStake);
+        }
+
+        // Decrease old delegate's delegated amount
+        delegators[_delegate].delegatedAmount = delegators[_delegate].delegatedAmount.sub(_amount);
+    }
+
+    /**
+     * @dev Tries to add a transcoder to active transcoder pool, evicts the active transcoder with the lowest stake if the pool is full
+     * @param _transcoder The transcoder to insert into the transcoder pool
+     * @param _totalStake The total stake for '_transcoder'
+     * @param _activationRound The round in which the transcoder should become active
+     */
+    function tryToJoinActiveSet(address _transcoder, uint256 _totalStake, uint256 _activationRound) internal {
+        uint256 pendingNextRoundTotalActiveStake = nextRoundTotalActiveStake;
+
+        if (transcoderPool.isFull()) {
+            address lastTranscoder = transcoderPool.getLast();
+            uint256 lastStake = transcoderTotalStake(lastTranscoder);
+
+            // If the pool is full and the transcoder has less stake than the least stake transcoder in the pool
+            // then the transcoder is unable to join the active set for the next round
+            if (_totalStake <= lastStake) {
+                return;
+            }
+
+            // Evict the least stake transcoder from the active set for the next round
+            // Not zeroing 'Transcoder.lastActiveStakeUpdateRound' saves gas (5k when transcoder is evicted and 20k when transcoder is reinserted)
+            // There should be no side-effects as long as the value is properly updated on stake updates
+            // Not zeroing the stake on the current round's 'EarningsPool' saves gas and should have no side effects as long as
+            // 'EarningsPool.setStake()' is called whenever a transcoder becomes active again.
+            transcoderPool.remove(lastTranscoder);
+            transcoders[lastTranscoder].deactivationRound = _activationRound;
+            pendingNextRoundTotalActiveStake = pendingNextRoundTotalActiveStake.sub(lastStake);
+
+            emit TranscoderEvicted(lastTranscoder);
+        }
+
+        transcoderPool.insert(_transcoder, _totalStake, address(0), address(0));
+        pendingNextRoundTotalActiveStake = pendingNextRoundTotalActiveStake.add(_totalStake);
+        Transcoder storage t = transcoders[_transcoder];
+        t.lastActiveStakeUpdateRound = _activationRound;
+        t.activationRound = _activationRound;
+        t.deactivationRound = 2**256 - 1;
```

Comment:
Can we make `2**256 - 1` a constant i.e. `MAX_FUTURE_ROUND`? Declaring this value as a constant can allow the compiler to compute the value at compile time so that the computation doesn't need to be done each time we assign the value to `deactivationRound`.

Reply:
Ah I see 

> If exponent is 0, gas used is 10. If exponent is greater than 0, gas used is 10 plus 10 times a factor related to how large the log of the exponent is.


So the `EXP` opcode might result in a higher gas cost here than `SLOAD`. The table is unclear on  the _factor_  but even if we assume a factor of `0.1` that would result in a gas cost of `10 + 256` which is higher than `200` from `SLOAD`.

Reply:
fixed in 7fddb5f